Meta/gn/build/BUILD.gn

```diff
@@ -154,10 +155,11 @@ config("compiler_defaults") {
   # Deterministic build setup, see
   # http://blog.llvm.org/2019/11/deterministic-builds-with-clang-and-lld.html
   if (is_clang) {
-    cflags += [
-      "-no-canonical-prefixes",
-      "-Werror=date-time",
-    ]
+    # FIXME: Clang from swiftly can't find stddef.h with this enabled.
```

Comment:
https://www.youtube.com/watch?v=RBGW9J3sx1g

Reply:
```
andrew@Andrew-Workstation:~/ladybird-org/ladybird-browser$ ccache clang++ --version
clang version 17.0.0 (https://github.com/swiftlang/llvm-project.git cdc8e8b92cfb4827f1f8fabd08d1db1e91498033)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/andrew/.local/bin
andrew@Andrew-Workstation:~/ladybird-org/ladybird-browser$ ninja -C out LibGfx
ninja: Entering directory `out'
[1/235] CXX obj/AK/AK.StackInfo.o
FAILED: obj/AK/AK.StackInfo.o 
ccache clang++ -MMD -MF obj/AK/AK.StackInfo.o.d -o obj/AK/AK.StackInfo.o -c ../AK/StackInfo.cpp  -I.. -Igen -Ivcpkg_installed/x64-linux/include -g1 -O2 -D_FILE_OFFSET_BITS=64 -DENABLE_COMPILETIME_FORMAT_CHECK -fdiagnostics-color -ffp-contract=off -fsigned-char -Wall -Wextra -Werror -Wno-invalid-offsetof -Wno-unknown-warning-option -Wno-unused-parameter -Wdelete-non-virtual-dtor -Wstring-conversion -Wno-user-defined-literals -fconstexpr-steps=16777216 -Wno-unused-private-field -Wno-implicit-const-int-float-conversion -Wno-vla-cxx-extension -Wno-unqualified-std-cast-call -no-canonical-prefixes -Werror=date-time -fno-semantic-interposition -std=c++23 -fvisibility-inlines-hidden -fno-exceptions -fPIC
In file included from ../AK/StackInfo.cpp:8:
In file included from ../AK/Platform.h:273:
/usr/include/unistd.h:226:10: fatal error: 'stddef.h' file not found
  226 | #include <stddef.h>
      |          ^~~~~~~~~~
1 error generated.
```

removing ` -no-canonical-prefixes` lets the build succeed.

```
andrew@Andrew-Workstation:~/ladybird-org/ladybird-browser/out$ ccache clang++ -MMD -MF obj/AK/AK.StackInfo.o.d -o obj/AK/AK.StackInfo.o -c ../AK/StackInfo.cpp  -I.. -Igen -Ivcpkg_installed/x64-linux/include -g1 -O2 -D_FILE_OFFSET_BITS=64 -DENABLE_COMPILETIME_FORMAT_CHECK -fdiagnostics-color -ffp-contract=off -fsigned-char -Wall -Wextra -Werror -Wno-invalid-offsetof -Wno-unknown-warning-option -Wno-unused-parameter -Wdelete-non-virtual-dtor -Wstring-conversion -Wno-user-defined-literals -fconstexpr-steps=16777216 -Wno-unused-private-field -Wno-implicit-const-int-float-conversion -Wno-vla-cxx-extension -Wno-unqualified-std-cast-call -Werror=date-time -fno-semantic-interposition -std=c++23 -fvisibility-inlines-hidden -fno-exceptions -fPIC
andrew@Andrew-Workstation:~/ladybird-org/ladybird-browser/out$ echo $?
0
```

Reply:
Perhaps something is funky with the toolchain shipped via `swiftly install main-snapshot-2024-09-18` + `swiftly use main-snapshot-2024-09-18` ? They could be building their clang dirs in a weird way.

Reply:
Oooh non-xcode clangs need `-isysroot $(xcrun -show-sdk-path)`. Maybe it's just that?

Reply:
…but stddef.h is a clang header, not a sysroot header, nvm.

Reply:
What do the `clang` symlinks point to?

Reply:
On linux, swiftly creates symlinks like so:

```
lrwxrwxrwx  1 andrew andrew   83 Sep 21 14:18 clang -> /home/andrew/.local/share/swiftly/toolchains/main-snapshot-2024-09-18/usr/bin/clang
lrwxrwxrwx  1 andrew andrew   85 Sep 21 14:18 clang++ -> /home/andrew/.local/share/swiftly/toolchains/main-snapshot-2024-09-18/usr/bin/clang++
lrwxrwxrwx  1 andrew andrew   86 Sep 21 14:18 clang-17 -> /home/andrew/.local/share/swiftly/toolchains/main-snapshot-2024-09-18/usr/bin/clang-17
lrwxrwxrwx  1 andrew andrew   89 Sep 21 14:18 clang-cache -> /home/andrew/.local/share/swiftly/toolchains/main-snapshot-2024-09-18/usr/bin/clang-cache
lrwxrwxrwx  1 andrew andrew   86 Sep 21 14:18 clang-cl -> /home/andrew/.local/share/swiftly/toolchains/main-snapshot-2024-09-18/usr/bin/clang-cl
lrwxrwxrwx  1 andrew andrew   87 Sep 21 14:18 clang-cpp -> /home/andrew/.local/share/swiftly/toolchains/main-snapshot-2024-09-18/usr/bin/clang-cpp
lrwxrwxrwx  1 andrew andrew   84 Sep 21 14:18 clangd -> /home/andrew/.local/share/swiftly/toolchains/main-snapshot-2024-09-18/usr/bin/clangd
```

... for like, every clang tool + swift tool + llvm tools. in ~/.local/bin, into `~/.local/share/swiftly/toolchains/<selected-toolchain>/usr`

Reply:
```
% cat test.c
#include <stddef.h>
% out/gn/bin/clang -c test.c
% ln -sf $PWD/out/gn/bin/clang clangtest
% ./clangtest -c test.c
% ./clangtest -no-canonical-prefixes -c test.c
test.c:1:10: fatal error: 'stddef.h' file not found
    1 | #include <stddef.h>
      |          ^~~~~~~~~~
1 error generated.
% out/gn/bin/clang -no-canonical-prefixes -c test.c
```

That's unfortunate.

Do you have to use those symlinks?

Reply:
I don't have to, I could use the realpath of them. but when we start messing around with more swift, it'll be annoying to have to say "btw, if you want to use gn, make sure to realpath your compilers" :shrug: